### ISSUE TITLE
NodaTime negative impact on serialisation perf

### ISSUE THREAD

**Issue by jskeet** [nodatime/nodatime.serialization#12]
As [raised in the nodatime repo](https://github.com/nodatime/nodatime/issues/844)

----

I'm not trying to be a pain about perf, but I've seen NodaTime appear in some new profile results, and on investigation, it seems that using `ConfigureForNodaTime` settings has a significant impact on Json.NET serialisation performance, *even when you're not serialising NodaTime types*.

I have a simple benchmark that does some variants of 
`JsonConvert.SerializeObject(new { tm = _instant }, _settings);` 

Sometimes _tm is set to a `DateTime` rather than a `Instant`, and I either do or don't include `_settings` in the call.

These are the results:
```
                           Method |       Mean |    Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
--------------------------------- |-----------:|---------:|----------:|-------:|---------:|-------:|----------:|
 SerializeDateTimeWithoutSettings |   890.2 ns | 11.56 ns |  10.81 ns |   1.00 |     0.00 | 0.3605 |   1.11 KB |
  SerializeInstantWithOutSettings | 1,719.1 ns | 34.18 ns |  33.57 ns |   1.93 |     0.04 | 0.5093 |   1.57 KB |
    SerializeDateTimeWithSettings | 3,286.9 ns | 64.51 ns | 104.17 ns |   3.69 |     0.12 | 0.4196 |    1.3 KB |
     SerializeInstantWithSettings | 2,863.0 ns | 57.10 ns |  72.21 ns |   3.22 |     0.09 | 0.5226 |   1.62 KB |
```
Obviously these aren't all serialising to the same text: the two `DateTime` versions are the same, but `Instant` only serialises properly if you include `_settings` - without it you get the members of Duration separately serialised.

It seems pretty obvious from the DotTrace picture below what the problem is - `NodaConverterBase` needs to find a smarter way to answer the `CanConvert` question than reaching for the reflection sledgehammer.  The obvious thing feels like a static dictionary cache of results, but one would need to think a bit about concurrency and whether there was some pathological bazillion-types scenario which could cause it to explode in length.

I think this is quite important because of the way that it impacts non NodaTime types so badly.

![nt2](https://user-images.githubusercontent.com/1306085/26902965-296dfc78-4bd3-11e7-9ade-1cc0beefa924.PNG)

**Comment by jskeet** [comment]
This has been raised in a PR already: #9. Time to add benchmarks in this repo and see what we can do.

@willdean: If I create a benchmark project, could you create a PR to add the benchmarks you quote above?

**Comment by stevenguh** [comment]
I am wondering instead of adding `JsonConverter`s in `JsonSerializer`, will annotating `JsonConverterAttribute` in runtime by `TypeDescriptor.AddAttributes` help with the perf issue as JamesNK mentioned the use of `JsonConverterAttribute`.

**Comment by jskeet** [comment]
I don't have the type to investigate that, but if you do, I'd be very interested to hear the results.

**Comment by stevenguh** [comment]
Using `TypeDescriptor` with `JsonConverterAttribute` does indeed improve pref from a preliminary benchmark result.

This is the code I used to benchmark.
https://gist.github.com/stevenguh/870e2c3fbc5af4ae44c7e45045e9b971

## TypeDescriptorBenchmarks
 |                           Method |          Mean |     StdDev | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
 |--------------------------------- |-------------- |----------- |------- |-------------- |------- |---------- |
 | SerializeDateTimeWithoutSettings |   951.5309 ns |  3.6833 ns |   1.00 |          0.00 | 0.3044 |   1.46 kB |
 |  SerializeInstantWithoutSettings | 2,503.6722 ns | 16.1143 ns |   2.63 |          0.02 | 0.4486 |   2.25 kB |
 |    SerializeDateTimeWithSettings | 1,016.2186 ns |  5.9504 ns |   1.07 |          0.01 | 0.3014 |   1.46 kB |
 |     SerializeInstantWithSettings | 2,496.1234 ns |  7.1323 ns |   2.62 |          0.01 | 0.4344 |   2.25 kB |

## ConfigureForNodaTimeBenchmarks
 |                           Method |          Mean |     StdDev | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
 |--------------------------------- |-------------- |----------- |------- |-------------- |------- |---------- |
 | SerializeDateTimeWithoutSettings |   955.8227 ns | 10.9985 ns |   1.00 |          0.00 | 0.3044 |   1.46 kB |
 |  SerializeInstantWithoutSettings | 2,393.0664 ns |  6.7800 ns |   2.50 |          0.03 | 0.4486 |   2.25 kB |
 |    SerializeDateTimeWithSettings | 2,854.1703 ns | 13.0733 ns |   2.99 |          0.04 | 0.3494 |   1.83 kB |
 |     SerializeInstantWithSettings | 3,071.8191 ns |  6.3313 ns |   3.21 |          0.04 | 0.3713 |   2.28 kB |

**Comment by jskeet** [comment]
Hmm. That may suggest it's worth doing in the JsonNet repo after all. Tricky.

**Comment by stevenguh** [comment]
Maybe I didn't follow all the discussion, what are you suggesting that should be done in the Json.NET repo?

**Comment by jskeet** [comment]
Sorry, that was very badly put - I mean in the serialization repo, in the JsonNet package.

**Comment by jskeet** [comment]
Another option to consider - using a contract resolver. See https://github.com/google/google-api-dotnet-client/pull/1080 for an example.

**Comment by stevenguh** [comment]
I am not sure how much PR would help in comparison to annotate a TypeConverter or JsonConverter on the Type. I can see it will cached the converter, but the resolver still need to go through a water fall type checking with CanConvert. If you have 15 converters cached in the contract resolver for example and one might still need to check 15 times of different converter's CanConvert method.

A side note, I realized once I set the TypeConverter on NodaTime Type, I don't really need the set up the JsonConverter for NodaTime types in the serializer as described in https://www.newtonsoft.com/json/help/html/SerializationGuide.htm
Basically, I don't need to add all the converters with ConfigureForNodaTime if I setup the TypeConverter correctly.

**Comment by jskeet** [comment]
@stevenguh: My impression was that the ContractResolver would be consulted once per *type* rather than on every value. I could be wrong.

Just to clarify, when you talk about TypeConverter, do you mean TypeDescriptor? (This is an area I know very little about, so I get easily confused.)

**Comment by stevenguh** [comment]
You are correct about the caching of the Converter if a custom contract resolver is used. Json.NET will cache the converter once it knew what converter to use after the first time from the custom contract resolver.

When I talked about the `TypeConverter` I was referring to annotating NodaTime time type statically (which is not possible without modifying the repo), or using `TypeDescriptor`.
If the `TypeConverter` is annotated on NodaTime, setting `JsonConverter` in `JsonSerializerSettings` or annotating `JsonConverter` with `JsonConverterAttribute` statically or via  `TypeDescriptor` is not needed expect for `Interval`.
`Interval` cannot serialize as dictionary key because the serialized JSON is a complex object as defined in `NodaIntervalConverter` instead of just a string . If we change that serialization then we will be able to serialize `Interval` as dictionary key with Json.NET, but that will be a breaking change.

I created a new benchmarks for a custom ContractResolver just like the in the CR you referenced, and also reran the benchmark above. Those didn't change much, so I not going to post them here again.
https://gist.github.com/stevenguh/7c10bbb16ed9d31897dc1b43c768005a

## NodaContractResolverBenchmarks
 |                           Method |          Mean |     StdDev | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
 |--------------------------------- |-------------- |----------- |------- |-------------- |------- |---------- |
 | SerializeDateTimeWithoutSettings |   986.7282 ns |  8.5615 ns |   1.00 |          0.00 | 0.3034 |   1.46 kB |
 |  SerializeInstantWithoutSettings | 2,627.5555 ns | 64.8934 ns |   2.66 |          0.07 | 0.4445 |   2.25 kB |
 |    SerializeDateTimeWithSettings | 1,078.1578 ns | 15.8540 ns |   1.09 |          0.02 | 0.3024 |   1.46 kB |
 |     SerializeInstantWithSettings | 1,748.1947 ns | 21.5579 ns |   1.77 |          0.03 | 0.3667 |   1.91 kB |

Apparently, using Contract Resolve is slightly faster then using `TypeDescriptor` to put `JsonConverter` via `JsonConverterAttribute` on the NodaType types.

But one problem is that if some one is using [CamelCasePropertyNamesContractResolver](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Serialization_CamelCasePropertyNamesContractResolver.htm)
1) They will somehow need to embed the `NodaTimeJsonContractResolve` (if we are proposing a contract resolver class for other people to use) because there is no multi-inheritance in c#
2) We can provide a utility type method to get the correct type of the converter base on a object type, so   a user is free from using any ContractResolver they have written. They will just need to override `JsonContract CreateContract(Type objectType)` and use the utility method there.

However, all these won't help with the serialization of NodaTime type as dictionary key. Serialization of the dictionary key will require TypeConverter being annotated on the type statically or by using `TypeDescriptor`

**Comment by jskeet** [comment]
Okay - I'd rather have one solution if possible. It sounds like TypeDescriptor/TypeConverter is a generally better approach, in terms of not worrying about the casing aspect, and also it handling the dictionary key parts.

So I think I'd like to just do the TypeConverter/TypeDescriptor approach. Does that sound right?

**Comment by stevenguh** [comment]
I agreed, I will prefer to have one solution. `TypeDescriptor`/`TypeConverter` seems like a solution that will work for all except `Interval`.

**Comment by jskeet** [comment]
I'm going to close this as it's had no further conversation in over 2 years - and in Noda Time 3 the types have type converters anyway.
